File: python/T215.py
from typing import List


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        return self.find(nums, len(nums) - k, 0, len(nums) - 1)

    def find(self, nums: List[int], k: int, start: int, end: int):
        if start == end:
            return nums[start]
        base = self.partition(nums, start, end)
        if base == k:
            return nums[k]
        elif base < k:
            return self.find(nums, k, base + 1, end)
        elif base > k:
            return self.find(nums, k, start, base - 1)
        else:
            return None

    # 分区函数，返回基准位置p, 基准位置左边的数字都<= pv， 基准位置右边的数字都 >= pv
    def partition(self, nums: List[int], start: int, end: int) -> int:
        p = nums[start]
        while start < end:
            while nums[end] >= p and start < end:
                end = end - 1
            nums[start] = nums[end]
            while nums[start] <= p and start < end:
                start = start + 1
            nums[end] = nums[start]

        nums[start] = p
        return start


if __name__ == '__main__':
    s = Solution()
    nums = [2, 1]
    k = 2
    result = s.findKthLargest(nums, k)
    print(result)

File: java/T215.java
package cc.peihan.java.all.leetcode;

public class T215 {

    public int findKthLargest(int[] nums, int k) {
        return quick_select(nums, nums.length - k, 0, nums.length - 1);
    }


    private int quick_select(int[] nums, int k, int start, int end) {
        if (start == end) {
            return nums[start];
        }
        int base = patition(nums, start, end);
        if (base == k) {
            return nums[base];
        } else if (base < k) {
            return quick_select(nums, k, base + 1, end);
        } else {
            return quick_select(nums, k, start, base - 1);
        }
    }

    private int patition(int[] nums, int start, int end) {
        int p = nums[start];
        while (start < end) {
            while (nums[end] >= p && start < end) {
                end--;
            }
            nums[start] = nums[end];
            while (nums[start] <= p && start < end) {
                start++;
            }
            nums[end] = nums[start];
        }
        nums[start] = p;
        return start;
    }


}
